### ISSUE TITLE
How to train own datasets?

### ISSUE THREAD

**Issue by zhanghan328** [amdegroot/ssd.pytorch#16]
I have own datasets with labeled. How to train it?

**Comment by amdegroot** [comment]
That is high on the to-do list but haven't gotten around to making that easy yet.  For now, I would recommend extending pytorch's dataset class so that it follows the functionality of our current VOC dataset class. Finishing up school next week and then I should be able to add that pretty quickly!

**Comment by zhanghan328** [comment]
Thank you for your reply. I have made my dataset format like VOC datasets. My dataset only has one class, is a hand in a image. I believe i made my dataset correctly. So I edited the num_classes = 2,net = build_ssd('train', 300, 2). When I trained it, I got the answer
![2017-05-08 13-00-48](https://cloud.githubusercontent.com/assets/16316669/25790924/5de2a6ee-33ef-11e7-9198-7880a2454629.png)
Maybe I would modify the multiboxloss. But i don't know how or how to fix it. Could you help me?

**Comment by andfoy** [comment]
I've actually tried to implement my own dataset according to the VOC dataloader. However at training time, I got NaN values at loss

**Comment by alexiskattan** [comment]
I'm also very interested in this.  @andfoy @amdegroot @zhanghan328  have working code or places to start?

**Comment by andfoy** [comment]
@alexiskattan I've actually successfully loaded Visual Genome dataset: https://github.com/andfoy/textobjdetection/blob/master/visual_genome_loader.py#L482

**Comment by ellisbrown** [comment]
here is our VOC Detection dataset: https://github.com/amdegroot/ssd.pytorch/blob/master/data/voc0712.py#L84

I would give it a look over, we have it extensively commented.  I largely based the dataset off of the datasets found here in pytorch's torchvision dataset repo: https://github.com/pytorch/vision/tree/master/torchvision/datasets

Sorry for the late responses -- Max and I are both currently travelling post graduation, but should be able to pick back up our slack here in a couple of weeks!

**Comment by alexiskattan** [comment]
Thank you you @ellisbrown and @andfoy.  I'm looking at it now.  I'm kinda new to this but will see how far I can get! <3

**Comment by igrekun** [comment]
Can confirm VGG gives NaN's with 2 class detection problem. Sometimes gives illegal memory access, which seems to be related to pytorch bug.

My MobileNet implementation achieved ~50% mAP on VOC2007 (160x160 input size).
Can PR it here after sorting out issues with HollywoodHeads dataset.

Seems like with smaller object priors are completely off and network doesn't train.

**Comment by alexiskattan** [comment]
@ellisbrown one thing that made training really easy for another library was to be able to input a text file that was like:

filepath,x1,y1,x2,y2,class_name
filepath,x1,y1,x2,y2,class_name
filepath,x1,y1,x2,y2,class_name

**Comment by alexiskattan** [comment]
@igrekun I'd be great to see how you did it.  Would you mind pointing to link where it is?

**Comment by amdegroot** [comment]
Until we get back from our summer vacations, I would recommend reading the issues people have posted on the original Caffè SSD repo regarding training custom datasets as I think I remember them being helpful.

**Comment by igrekun** [comment]
Just returned from my mountain trip and have to sort out work-related issues, so here is dirty and hacky gist to train mobilenet + SSD.

https://gist.github.com/igrekun/35129c5a1d4f7ef2ffecdf0560adfc1f

You can download pretrained on ImageNet MobileNet head here: 
https://s3.us-east-2.amazonaws.com/ig-models/mobilenet/mbnethead.pth

The trained model with 50.22 mAP on pascal is stored here:
https://s3.us-east-2.amazonaws.com/ig-models/mobilenet/mbnet_iter_59800.pth

I'm busy for the next few days, but if you are interested, I will file a proper pull request.
There is some redundancy with not using cfg values to initialize SSD prior boxes but we can sort that out.

Otherwise the codebase is clean and a pleasure to work with. Thanks, Max! (:

**Comment by jtoy** [comment]
@andfoy thanks for sharing your training implementation. What is the link for the article "Human-related Object Detection based on Natural Language Parsing of Image Query Expressions article"? is it this? http://ieeexplore.ieee.org/document/5363983/ cant seem to find it.

**Comment by andfoy** [comment]
@jtoy I'm actually on process of writing it, so is not published yet!

**Comment by romaniukm** [comment]
@igrekun do you know what PyTorch bug this is? I've been getting strange errors while training to detect a single class of objects. But it occurs all the time from the start so could be an entirely different problem (or indeed some bug in my own code).
> Can confirm VGG gives NaN's with 2 class detection problem. Sometimes gives illegal memory access, which seems to be related to pytorch bug.

**Comment by amdegroot** [comment]
@igrekun I apologize for my tardy response.  Feel free to submit a PR!  I'm just getting back from a long summer break myself.

**Comment by aveysov** [comment]
Check out my solution here - https://github.com/amdegroot/ssd.pytorch/issues/72

**Comment by Kfirbarzilay** [comment]
Hi,
@alexiskattan how can I install visual genome?
thanks :)

**Comment by li10141110** [comment]
@zhanghan328 hi,could you please tell me how's your work going on,I have the same problem as yours. My wechat(微信)：lijingx-    thank you in advance!

**Comment by dongzhuoyao** [comment]
any advance?

**Comment by shreyank13** [comment]
Hey guys, interested to know if there was any progress on this??

**Comment by wagnzi** [comment]
any progress?

**Comment by Ai-is-light** [comment]
I trained 12 categories dataset on this repo, firstly, I transformed my data to like coco format. And, I run the demo.py and can get normal results. however, when I want to use the coco api to eval my model for getting the mAP like coco dataset, I failed. But, I got all 0 result or nearly 0, which is likely the @banxia1994 and @Hliang1994
I got
Running per image evaluation...
Evaluate annotation type bbox
DONE (t=0.36s).
Accumulating evaluation results...
DONE (t=0.10s).
Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000


The following code as my reference:
I used the @lpd2csx code to record the ouput https://github.com/amdegroot/ssd.pytorch/issues/253

There are anyone eval the likely coco's format dataset' mAP results?
looking forward to any reply.

**Comment by lesonly** [comment]
> I trained 12 categories dataset on this repo, firstly, I transformed my data to like coco format. And, I run the demo.py and can get normal results. however, when I want to use the coco api to eval my model for getting the mAP like coco dataset, I failed. But, I got all 0 result or nearly 0, which is likely the @banxia1994 and @Hliang1994
> I got
> Running per image evaluation...
> Evaluate annotation type bbox
> DONE (t=0.36s).
> Accumulating evaluation results...
> DONE (t=0.10s).
> Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
> Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.000
> Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.000
> Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
> Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
> Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
> Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
> 
> The following code as my reference:
> I used the @lpd2csx code to record the ouput #253
> 
> There are anyone eval the likely coco's format dataset' mAP results?
> looking forward to any reply.

Happy new year!Have you solved this problem?

**Comment by Hemantr05** [comment]
> I've actually tried to implement my own dataset according to the VOC dataloader. However at training time, I got NaN values at loss

How may I train my custom coco dataset? @andfoy

**Comment by zzu0654** [comment]
> @ellisbrown one thing that made training really easy for another library was to be able to input a text file that was like:
> 
> filepath,x1,y1,x2,y2,class_name
> filepath,x1,y1,x2,y2,class_name
> filepath,x1,y1,x2,y2,class_name

My thoughts are the same as yours.Do you realize the loading of this data format？

**Comment by 201814476** [comment]
> > I trained 12 categories dataset on this repo, firstly, I transformed my data to like coco format. And, I run the demo.py and can get normal results. however, when I want to use the coco api to eval my model for getting the mAP like coco dataset, I failed. But, I got all 0 result or nearly 0, which is likely the @banxia1994 and @Hliang1994
> > I got
> > Running per image evaluation...
> > Evaluate annotation type bbox
> > DONE (t=0.36s).
> > Accumulating evaluation results...
> > DONE (t=0.10s).
> > Average Precision (AP) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
> > Average Precision (AP) @[ IoU=0.50 | area= all | maxDets=100 ] = 0.000
> > Average Precision (AP) @[ IoU=0.75 | area= all | maxDets=100 ] = 0.000
> > Average Precision (AP) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
> > Average Precision (AP) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
> > Average Precision (AP) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
> > Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 1 ] = 0.000
> > Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets= 10 ] = 0.000
> > Average Recall (AR) @[ IoU=0.50:0.95 | area= all | maxDets=100 ] = 0.000
> > Average Recall (AR) @[ IoU=0.50:0.95 | area= small | maxDets=100 ] = 0.000
> > Average Recall (AR) @[ IoU=0.50:0.95 | area=medium | maxDets=100 ] = 0.000
> > Average Recall (AR) @[ IoU=0.50:0.95 | area= large | maxDets=100 ] = 0.000
> > The following code as my reference:
> > I used the @lpd2csx code to record the ouput #253
> > There are anyone eval the likely coco's format dataset' mAP results?
> > looking forward to any reply.
> 
> Happy new year!Have you solved this problem?

hello，Have you solved this problem?

**Comment by MaxwellHogan** [comment]
Have anyone come up with a method for choosing the values for the default boxes in config.py - see below
```
    'min_sizes': [30, 60, 111, 162, 213, 264],
    'max_sizes': [60, 111, 162, 213, 264, 315],
    'aspect_ratios': [[2], [2, 3], [2, 3], [2, 3], [2], [2]],
```
